pkg/render/kube-controllers.go

```diff
@@ -54,55 +61,93 @@ func KubeControllers(
 	authentication *operator.Authentication,
 	enabledESOIDCWorkaround bool,
 	clusterDomain string,
-	esAdminSecret *v1.Secret,
+	kubeControllersUserSecret *v1.Secret,
+	kubeControllersVerificationSecret *v1.Secret,
+	esAdminUserSecret *v1.Secret,
 	metricsPort int,
 ) *kubeControllersComponent {
-	var elasticsearchUserSecret *v1.Secret
-	if esAdminSecret != nil {
-		elasticsearchUserSecret = &v1.Secret{
+	var kubeControllersSecureUserSecret *v1.Secret
+	if kubeControllersUserSecret == nil {
+		password := randomString(16)
+		kubeControllersUserSecret = &v1.Secret{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      ElasticsearchKubeControllersUserSecret,
 				Namespace: common.CalicoNamespace,
 			},
+			Data: map[string][]byte{
+				"username": []byte(ElasticsearchKubeControllersUserName),
+				"password": []byte(password),
+			},
+		}
+	}
+	// Kube controllers no longer uses admin credentials to make requests directly to Elasticsearch. Instead, dummy credentials are generated and stored in
+	// the user secret, a hashed version of the credentials is stored in the tigera-elasticsearch namespace for ES Gateway to retrieve and use to compare
+	// the dummy credentials, and a secret containing real admin level credentials is created and stored in the tigera-elasticsearch namespace to be swapped in once
+	// ES Gateway has confirmed that the dummy credentials match.
+	hashedUserSecret := kubeControllersVerificationSecret
+	if hashedUserSecret == nil {
+		hashedPassword, _ := bcrypt.GenerateFromPassword(kubeControllersUserSecret.Data["password"], bcrypt.MinCost)
+		hashedUserSecret = &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ElasticsearchKubeControllersVerificationUserSecret,
+				Namespace: ElasticsearchNamespace,
```

Comment:
If I get it right, the kube controllers repo will be changed to monitor this secret, rather than the tigera-secure-es-user (or something like that). If that assumption is correct:
- The secret used by kube controllers should be in the calico-system namespace
- The render function of kube controllers should not be responsible for creating anything in the elasticsearch namespace. (Is the existence of that namespace even guaranteed here?) Normally, it copies things to the tigera-operator namespace.
- We should restrict rbac such that it can no longer access secrets that it does not need.

Reply:
The ElasticsearchKubeControllersVerificationUserSecret stores the hashed password from ElasticsearchKubeControllersUserSecret and is used by ES gateway to authenticate the dummy credentials that kube-controllers uses in its requests. It needs to live in the elasticsearch namespace so that the gateway can access it. 

> If I get it right, the kube controllers repo will be changed to monitor this secret, rather than the tigera-secure-es-user.

Kube controllers will still use the same secret name as before, the difference is that it will contain dummy credentials now created in the lines above this comment. The admin level credentials will be stored in the Elasticsearch namespace and es gateway will swap them in once it authenticates the dummy credentials.

> The render function of kube controllers should not be responsible for creating anything in the elasticsearch namespace. (Is the existence of that namespace even guaranteed here?) Normally, it copies things to the tigera-operator namespace.

I'm not sure how to get around this, I guess one option would be to move this responsibility to the esgateway component? Have kube-controllers just generate the secret with the dummy credentials (ElasticsearchKubeControllersUserSecret) and then use that to create ElasticsearchKubeControllersVerificationUserSecret and ElasticsearchKubeControllersSecureUserSecret in esgateway rendering.

Reply:
@rene-dekker is right (I commented on this earlier as well), kube controllers shouldn't create anything in the elasticsearch namespace.

These secrets aren't generated by the core controller, there's no need to pass them down to be copied here. The Logstorage controller should be responsible for copying these secrets to the elasticseach namespace, the only secret this controller should be responsible for is the `ElasticsearchKubeControllersVerificationUserSecret`, because that's the only one it actually uses.